### PR TITLE
Tm 6488 add eltoml

### PR DESCRIPTION
**Pull request by mansi-mp-200**
Dual Merge: 
- [FE PR](https://github.com/MetaPhase-Consulting/State-TalentMAP/pull/3105)
- [Mock PR](https://github.com/MetaPhase-Consulting/mock-fsbid/pull/509)

[Ticket][(linkToTicket)](https://metaphase.atlassian.net/jira/software/c/projects/TM/boards/1?assignee=712020%3A3d9639a5-439c-4e2d-9bce-37b6924384d2&selectedIssue=TM-6488)

This endpoint is add the ELTOML key to every EL position object we get back from BackOfficeCRUD. Currently this functionality is in the SPRINT101 package but will be moved up to WEB API WRAP and change in our code as well once its done on WS end (this is marked by a comment).

### CODE REVIEW

File: talentmap_api/fsbid/services/common.py [hunk 1]
```diff
@@ -301,18 +302,25 @@ def get_individual(uri, query, query_mapping_function, jwt_token, mapping_functi
...
-
 # for calls to BackOffice CRUD POST EP
 def send_post_back_office(proc_name, package_name, request_body, request_mapping_function, response_mapping_function, jwt_token):
     url = f"{BACKOFFICE_CRUD_URL}?procName={proc_name}&packageName={package_name}"
-    json_body = request_mapping_function(request_body)
+    if request_mapping_function:
+        json_body = request_mapping_function(request_body)
+    else:
+        json_body = request_body
+    logger.info(f"FSBid backoffice call for procedure {proc_name} with body {json_body}")
```
Comment (M00staff) [inline talentmap_api/fsbid/services/common.py:312]: did you mean to keep this logger statement in?

i posted in the dev channel chat about how to test this locally

File: talentmap_api/fsbid/services/common.py [hunk 2]
```diff
@@ -301,18 +302,25 @@ def get_individual(uri, query, query_mapping_function, jwt_token, mapping_functi
     return pydash.get(response, '[0]') or None
 
 
-
 # for calls to BackOffice CRUD POST EP
 def send_post_back_office(proc_name, package_name, request_body, request_mapping_function, response_mapping_function, jwt_token):
     url = f"{BACKOFFICE_CRUD_URL}?procName={proc_name}&packageName={package_name}"
-    json_body = request_mapping_function(request_body)
+    if request_mapping_function:
```
Comment (M00staff) [inline talentmap_api/fsbid/services/common.py:308]: my understanding is that all of the backOffice calls are structured the same way and will have both a request mapping function and a response mapping function 

image

so is this if block necessary?

File: talentmap_api/fsbid/services/common.py [hunk 3]
```diff
@@ -301,18 +302,25 @@ def get_individual(uri, query, query_mapping_function, jwt_token, mapping_functi
...
         response = requests.post(url, headers={'JWTAuthorization': jwt_token, 'Content-Type': 'application/json'}, json=json_body).json()
     except:
         logger.error(f"FSBid backoffice call for procedure {proc_name} failed.")
         raise Exception('Error at FSBid call')
 
-    return response_mapping_function(response)
+    if response_mapping_function:
+        return response_mapping_function(response)
+    else:
+        return response
```
Comment (M00staff) [inline talentmap_api/fsbid/services/common.py:322]: makes sense to keep this one - if we want to just pass the response back to the frontend without any mapping 🤔

File: talentmap_api/fsbid/services/common.py [hunk 4]
```diff
@@ -26,6 +26,7 @@
 from talentmap_api.fsbid.services import employee as empservices
 from talentmap_api.fsbid.services import agenda as agendaservices
 from talentmap_api.fsbid.requests import requests
+# import requests # type: ignore
```
Comment (M00staff) [inline talentmap_api/fsbid/services/common.py:29]: did you mean to keep this in?

File: talentmap_api/fsbid/services/positions.py [hunk 5]
```diff
@@ -341,7 +343,11 @@ def el_postions_req_mapping(request):
     }
     for key in request:
         values_formatted = []
-        if key == 'el-tps':
+        if key == 'page':
+            result['PV_PAGE_I'] = request[key]
+        elif key == 'limit':
+            result['PV_PAGE_ROWS_I'] = request[key]
+        elif key == 'el-tps':
```
Comment (M00staff) [inline talentmap_api/fsbid/services/positions.py:350]: 👍

File: talentmap_api/fsbid/services/positions.py [hunk 6]
```diff
@@ -481,18 +496,33 @@ def languages_map(x):
...
-        "proc_name": "prc_iud_tracking_details_grid",
-        "package_name": "PKG_WEBAPI_WRAP_SPRINT101",
-        "request_body": {},
-        # "request_mapping_function": edit_el_positions_req_mapping,
-        # "response_mapping_function": edit_el_positions_res_mapping,
+        "proc_name": "prc_iud_tracking_details_pos",
+        "package_name": "PKG_WEBAPI_WRAP_SPRINT101", # should be changed to PKG_WEBAPI_WRAP in the future 
+        "request_body": json_input,
+        "request_mapping_function": None,
+        "response_mapping_function": None,
```
Comment (M00staff) [inline talentmap_api/fsbid/services/positions.py:519]: oh, i see ... 🤔

this is a slightly different pattern than what we've been using - if it works, it works and is fine, but you might want to check out these two files to see the pattern that we've been using with the backOffice endpoints and try to stick to that going forward. 

`talentmap_api/fsbid/services/bid_audit.py`
`talentmap_api/fsbid/services/assignment_cycles.py`

it'll just make it easier on everyone

File: talentmap_api/fsbid/urls/positions.py [hunk 7]
```diff
@@ -11,7 +11,8 @@
     url(r'^frequent_positions/$', views.FSBidFrequentPositionsView.as_view(), name="FSBid-frequent-positions"),
     url(r'^el_positions/$', views.FSBidEntryLevelPositionsView.as_view(), name="FSBid-el-positions"),
     url(r'^el_positions/filters/$', views.FSBidEntryLevelPositionsFiltersView.as_view(), name='FSBid-el-positions-filters'),
-    url(r'^el_positions/save/(?P<pk>[0-9]+)/$', views.FSBidEntryLevelPositionsActionView.as_view(), name='FSBid-el-positions-edit'),
+    # url(r'^el_positions/save/(?P<pk>[0-9]+)/$', views.FSBidEntryLevelPositionsActionView.as_view(), name='FSBid-el-positions-edit'),
```
Comment (M00staff) [inline talentmap_api/fsbid/urls/positions.py:14]: another commented out line - did we want to keep this in?

**Review by M00staff — APPROVED** [review]
if this is working, then I'm not going to hold it up - but def check out these two files to get a feel for how we build out the BackOffice requests

`talentmap_api/fsbid/services/bid_audit.py`
`talentmap_api/fsbid/services/assignment_cycles.py`

there are others, and we've been pretty uniform on how they all look - will make bug fixes easier to work with ... if you have the time and want to clean it up, knock yourself out - but I'm not going to hold this up 👍

File: talentmap_api/fsbid/services/common.py (hunk 1 above)
Comment (mansi-mp-200) [reply talentmap_api/fsbid/services/common.py:312]: Oh I forgot to take it out I think. Good catch, I'll clean it up. Thank. you

File: talentmap_api/fsbid/services/positions.py [hunk 8]
```diff
@@ -481,18 +496,34 @@ def languages_map(x):
...
+
+    # Web Service JSON Input for EL Position edit
+    payload = {
+                    "PV_API_VERSION_I": "",
+                    "PV_AD_ID_I": "",
+                    "PV_ACTION_I": "",
+                    "PTYP_CUST_TD_POS_TAB_I": {"Data": data}
+                }
+    # convert payload to json string
+    json_input = json.dumps(payload)
```
Comment (M00staff) [inline talentmap_api/fsbid/services/positions.py:513]: this is nifty!

yea - some of the requests have to be structured in really, really weird ways

File: talentmap_api/fsbid/views/positions.py [hunk 9]
```diff
@@ -81,11 +81,13 @@ def get(self, request):
         return Response(services.get_el_positions_filters(request.query_params, request.META['HTTP_JWT']))
 
 class FSBidEntryLevelPositionsActionView(BaseView):
-    permission_classes = (IsAuthenticatedOrReadOnly, isDjangoGroupMember('superuser'))
     '''
     Edit and save an EL Position
     '''
+    permission_classes = (IsAuthenticatedOrReadOnly, isDjangoGroupMember('superuser'))
```
Comment (M00staff) [inline talentmap_api/fsbid/views/positions.py:87]: i think we'll eventually need to capture the `el_cdo` role here - I'm not sure how its all wired up off the top of my head - but we use this to enforce roles for the API and they hit and unauthorized error if they do not have the correct role 

image

**Pull request merged**